api/src/main/java/org/apache/iceberg/catalog/CatalogTransaction.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.catalog;
+
+public interface CatalogTransaction {
+
+  enum IsolationLevel {
+
+    /**
+     * Table can change underneath but still have to pass normal table validation. Loading table
+     * data when it's referenced.
+     */
+    SNAPSHOT,
+
+    /**
+     * All tables participating in the transaction must be in the same state when committing
+     * compared to when the transaction started. Loading table data when it is referenced.
+     */
+    SERIALIZABLE_WITH_LOADING_READS,
+
+    /**
+     * All tables participating in the transaction must be in the same state when committing
+     * compared to when the transaction started. Loading table data as of TX start time.
+     */
+    SERIALIZABLE_WITH_FIXED_READS,
+  }
```

Comment:
I know this is still draft, but I wanted to verify for multi-table transactions shouldn't the isolation level enforcement be done at the branch level and not just the table level?

For example, for serializable with fixed reads I would expect the validation that the branch is in the same state even if another branch is written to.

```
    CatalogTransaction catalogTx = catalog.startTransaction(isolationLevel);
    Catalog txCatalog = catalogTx.asCatalog();
    catalog.loadTable(table1).newFastAppend().appendFile(FILE_B).toBranch("branch1").commit();
    catalog.loadTable(table2).newFastAppend().appendFile(FILE_C).toBranch("branch2").commit();
```

Let's say the append to table 1 above happens concurrently with an UPDATE/DELETE which matches but the update/delete happens on "branch3", then the multi-table transaction should still succeed.

Reply:
it should actually be most likely at the branch level (but I'd have to revisit and think about this). However, I started to work on this before writing to branches was committed, hence it's not being reflected here.